### PR TITLE
[WIP]faster histogram sum up, also better accuracy. 

### PR DESCRIPTION
**Pull request by guolinke**
comparison:

speed (average of run 3 times):

|  Data | master branch | this PR |
|----|  ----| ---- |
| Higgs|238 s |  232s |
| Yahoo LTR |163 s | 140 s|
| MS LTR |227 s | 212s |

Accuracy:

Higgs's AUC:

| Metric      |master branch | This PR|
|----|  ----| ---- | 
| AUC|0.845612 | 0.845696|

ndcg at Yahoo LTR:

| Metric      |master branch | This PR |
|----|  ----| ---- | 
| ndcg@1 |0.733789|0.734298|
| ndcg@3 |0.74039 |0.739005|
| ndcg@5 |0.757809|0.756383|
| ndcg@10|0.797109|0.796678|

ndcg at MS LTR:

| Metric      | master branch |This PR |
|----|  ----| ---- | 
| ndcg@1 |0.523594|0.524338|
| ndcg@3 |0.504672|0.505648|
| ndcg@5 |0.510191|0.510115|
| ndcg@10|0.527902|0.527797|

### CODE REVIEW

**Comment by gugatr0n1c** [comment]
@guolinke can you please provide more anfo about "also better acurracy"?
tnx!

File: CMakeLists.txt [hunk 1]
```diff
@@ -47,7 +47,7 @@ if(USE_GPU)
 endif(USE_GPU)
 
 if(UNIX OR MINGW OR CYGWIN)
-    SET(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -pthread -O3 -Wall -std=c++11 -Wno-ignored-attributes")
+    SET(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -pthread -O3 -Wall -std=c++11 -Wno-ignored-attributes -march=native")
```
Comment (Laurae2) [inline CMakeLists.txt]: @guolinke Perhaps `-mtune=native` is better for those using portable installations. We could do this:

* `-march=core2` so anyone benefits from core2 instructions
* `-mtune=native` so it is tailored for the current CPU used during compilation, and fallbacks to core2 if native instructions are not available

**Comment by guolinke** [comment]
@gugatr0n1c 
The previous code have some sum-up issues related to float point. 
consider following code:
```
double s = 0.0f;
for(int i = 0; i < n ; ++i) s += a[i];
```
when ```i``` become larger, the ```s``` is far large than ```a[i]```. Thus, the sum-up will be not correct.

[Kahan summation](https://en.wikipedia.org/wiki/Kahan_summation_algorithm) can solve this problem. But it have some additional cost. 

So i use a simple solution here:

```
double s = 0.0f;
int i = 0;
while(i < n){
  double ts = 0.0f;
  for(int j= 0; j < 65536 && i < n; ++j, ++i){
    ts += a[i];
  }
  s += ts;
}
```

```ts``` is the sum-up of 65536 element, so add its value to ```s``` will be more accuracy. 

I will paste the benchmark result soon.

File: CMakeLists.txt (hunk 1 above)
Comment (guolinke) [inline CMakeLists.txt]: Did you mean that add both ```-march=core2 -mtune=native``` ?

Comment (Laurae2) [inline CMakeLists.txt]: yes, allows portability between any >=core2 CPUs while keeping native performance

**Comment by guolinke** [comment]
@huanzhang12 
I change the type of histogram from ```double``` to ```float```. It may needs some changes for the GPU codes as well. And the parameter ```gpu_use_dp``` can be removed now.

**Pull request merged**